docs/CONFIG.md
# wbenbihi/hourglasstensorflow - Configuration

## mode

- REQUIRED
- Type: enum
  - `train`
  - `test`
  - `inference`
  - `server`
- Mode is used to define the behavior to adopt:
  - `train`: Train a model according to the [Train](#train) configuration
  - `test`: Launch inference on already labeled samples according to the [Test](#test) configuration
  - `inference`: Launch inference on given list of image and produce an aggregated table with prediction
  - `server`: Serve the model as an API

## data

`object`

- `OPTIONAL`
- Type: [`HTFObjectReference[BaseDataHandler]`](../hourglass_tensorflow/types/config/fields.py)
- Default: `hourglass_tensorflow.handlers.data.HTFDataHandler`
- Reference to the handler to use for the Data processing _(see [doc/HANDLERS](./HANDLERS.md))_

### data.input

`mode`

- `OPTIONAL`
- Type: [`Enum(RGB, BGR, GRAY, RGBA, BGRA):ImageModesType`](../hourglass_tensorflow/types/config/data.py)
- Default: `RGB`
- _(WIP NOT IMPLEMENTED)_ Image mode

`source`

- `REQUIRED (TRAIN|TEST|INFERENCE)`
- Type: `str`
- Folder containing images to use

`extensions`

- `REQUIRED`
- Type: `list[str]`
- File extensions to read in `$data.input.source`

### data.output

`source`

- `REQUIRED (TRAIN|TEST)`
- Type: `str`
- CSV file containing the labels. Only required for modes using labeled data _(Train & Test modes)_

`column_source`

- `REQUIRED (TRAIN|TEST)`
- Type: `str`
- Column name in `$data.output.source` containing the image path

`column_set`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Column name in `$data.output.source` containing the split set information _(Does the sample belong to TRAIN/TEST/VALIDATION)_. **Not required**

`source_prefixed`

- `OPTIONAL`
- Type: `bool`
- Default: `false`
- Does `$data.output.column_source` contains the folder `$data.input.source`. If `false` the **default** handler will prefix.

`prefix_columns`

- `OPTIONAL (TRAIN|TEST)`
- Type: `list[str]`
- List of additional columns' name available in `$data.output.source`

#### data.output.joints

`num`

- `OPTIONAL (TRAIN|TEST)`
- Type: `int`
- Default: `16 (MPII compliant)`
- Number of joints/keypoints referenced in the dataset

`naming_convention`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Default: `joint_{JOINT_ID}_{SUFFIX}`
- Naming convention for columns related to joints. _You can specify dynamic structures if you have additional information about the joints (e.g `joint_{JOINT_ID}_{SUFFIX}_{ADDITIONAL}`)_

`dynamic_fields`

- `OPTIONAL (TRAIN|TEST)`
- Type: `list[str]`
- Default: `["SUFFIX"]`
- The group name of fields referenced in the naming convention.Naming convention for columns related to joints. _If you specified the following naming convention `joint_{JOINT_ID}_{SUFFIX}_{ADDITIONAL}` you would have to specify dynamic fields as `["SUFFIX", "ADDITIONAL"]`_

`names`

- `OPTIONAL (TRAIN|TEST)`
- Type: `list[str]`
- Default: _See [train.default.yaml](../config/train.default.yaml)_
- List the joints/keypoints' name. **List length must be equal to `$data.output.joints.num`**

#### data.output.joints.format

`id_field`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Default: `JOINT_ID`
- Reference the part of `$data.output.joints.naming_convention` relative to the joint ID number

`$GROUP`

- `OPTIONAL (TRAIN|TEST)`
- Type: `dict`
- You can define the columns name of a given group from the `$data.output.joints.naming_convention`. _(See [train.default.yaml:SUFFIX](../config/train.default.yaml))_

## dataset

`object`

- `OPTIONAL`
- Type: [`HTFObjectReference[BaseDatasetHandler]`](../hourglass_tensorflow/types/config/fields.py)
- Default: `hourglass_tensorflow.handlers.dataset.HTFDatasetHandler`
- Reference to the handler to use for the Dataset generation _(see [doc/HANDLERS](./HANDLERS.md))_

`image_size`

- `OPTIONAL`
- Type: `int`
- Default: `256`
- The size of the squared image as tensor input. Will constraint all image from the [Data](#data) configuration to have a size of [`image_size`, `image_size`, `3`]
  - Bigger `image_size` will results in slower computation

`column_image`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Default: `image`
- The name of the column containing the image path from `$data.output.source`

`normalization`

- `OPTIONAL`
- Type: [`Enum(ByMax, L2, Normal, FromZero, AroundZero):NormalizationModeType`](../hourglass_tensorflow/types/config/dataset.py)
- Default: `image`
- Normalization to apply to images and heatmaps
  - `ByMax`: Will constraint the Value between 0-1 by dividing by the global maximum
  - `L2`: Will constraint the Value by dividing by the L2 Norm on each channel
  - `Normal`: Will apply (X - Mean) / StdDev**2 to follow normal distribution on each channel
  - `FromZero`: Origin is set to 0 maximum is 1 on each channel
  - `AroundZero`: Values are constrained between -1 and 1

`augmentation`

- `OPTIONAL`
- Type: [`HTFObjectReference`](../hourglass_tensorflow/types/config/fields.py)
- Default: `[]`
- Data augmentation mapper to apply to a tensorflow Dataset
- Reference to mapping function to use for the Dataset generation _(see [doc/HANDLERS](./HANDLERS.md))_

### dataset.heatmap

`channels`

- `OPTIONAL (TRAIN|TEST)`
- Type: `int`
- Default: `16`
- The number of heatmap channels. Usually equal to `$data.output.joints.num` as it represents the joint heatmap.

`size`

- `OPTIONAL (TRAIN|TEST)`
- Type: `int`
- Default: `64`
- The size of the squared heatmap. Will constraint all heatmaps generated to have a size of [`size`, `size`, `channels`]

`stacks`

- `OPTIONAL (TRAIN|TEST)`
- Type: `int`
- Default: `2`
- The number of stacks per heatmap. This field is used in conjonction with intermediate supervision

`stddev`

- `OPTIONAL (TRAIN|TEST)`
- Type: `float`
- Default: `5.`
- The standard deviation to use for the 2D heatmap generation.
  - **This is a model HYPERPARAMETER** as a bigger value will increase the convergence power of the model while decreasing its prediction precision.
  - The standard deviation constraint the region of interest to look for a given joint

### dataset.sets

`split_by_column`

- `OPTIONAL (TRAIN|TEST)`
- Type: `bool`
- Default: `true`
- True if you want to use `$dataset.sets.column_split` as the column from `$data.output.source` to split your dataset. False if you want to apply a random selection accross your samples in run.

`column_split`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Default: `set`
- Column name from `$data.output.source` containing the sets of the sample.

`value_train`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Default: `TRAIN`
- The value in column `$dataset.sets.column_split` from `$data.output.source` to identify **train** samples.  _Requires `$dataset.sets.split_by_column` to be `True`_

`value_test`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Default: `TEST`
- The value in column `$dataset.sets.column_split` from `$data.output.source` to identify **test** samples. _Requires `$dataset.sets.split_by_column` to be `True`_

`value_validation`

- `OPTIONAL (TRAIN|TEST)`
- Type: `str`
- Default: `VALIDATION`
- The value in column `$dataset.sets.column_split` from `$data.output.source` to identify **validation** samples. _Requires `$dataset.sets.split_by_column` to be `True`_

`train`

- `OPTIONAL (TRAIN|TEST)`
- Type: `bool`
- Default: `true`
- Samples a **train** set from the dataset. _Requires `$dataset.sets.split_by_column` to be `False`_

`test`

- `OPTIONAL (TRAIN|TEST)`
- Type: `bool`
- Default: `true`
- Samples a **test** set from the dataset. _Requires `$dataset.sets.split_by_column` to be `False`_

`validation`

- `OPTIONAL (TRAIN|TEST)`
- Type: `bool`
- Default: `true`
- Samples a **validation** set from the dataset. _Requires `$dataset.sets.split_by_column` to be `False`_

`ratio_train`

- `OPTIONAL (TRAIN|TEST)`
- Type: `float in ]0-1]`
- Default: `0.7`
- The ratio of **train** datapoints to sample from the unsplitted dataset. _Requires `$dataset.sets.split_by_column` to be `False`_

`ratio_test`

- `OPTIONAL (TRAIN|TEST)`
- Type: `float in ]0-1]`
- Default: `0.15`
- The ratio of **test** datapoints to sample from the unsplitted dataset. _Requires `$dataset.sets.split_by_column` to be `False`_

`ratio_validation`

- `OPTIONAL (TRAIN|TEST)`
- Type: `float in ]0-1]`
- Default: `0.15`
- The ratio of **validation** datapoints to sample from the unsplitted dataset. _Requires `$dataset.sets.split_by_column` to be `False`_

### dataset.bbox

`activate`

- `OPTIONAL (TRAIN|TEST)`
- Type: `bool`
- Default: `true`
- Will use the joint coordinates to compute a bounding box and crop the bounding box from the image.
  - `$dataset.bbox.activate` makes the model more precise by supplying only one person per sample. On a production environment, enabling this setting might make the model worst as it will encounter images with several person.

`factor`

- `OPTIONAL (TRAIN|TEST)`
- Type: `float`
- Default: `1.`
- Expansion factor for the bounding box

## model

`object`

- `OPTIONAL`
- Type: [`HTFObjectReference[BaseModelHandler]`](../hourglass_tensorflow/types/config/fields.py)
- Default: `hourglass_tensorflow.handlers.model.HTFModelHandler`
- Reference to the handler to use for the graph architecture generation _(see [doc/HANDLERS](./HANDLERS.md))_

`build_as_model`

- `DEBUG`
- `OPTIONAL`
- Type: `bool`
- Default: `true`
- When `true` will use generate a [HourglassModel](../hourglass_tensorflow/models/hourglass.py) instance. If `false` will recreate each layer and build a graph that is not encapsulated within a `keras.Model`.
  - **DEBUG** settings, do not set to `false` as serialization might not work properly.

`data_format`

- `DEBUG`
- `OPTIONAL`
- Type: `Enum(NHWC)`
- Default: `NHWC`
- The tensor format to use in the model. This model supports only `NHWC` as it is the default tensor format use by `tensorflow`

### model.params

`intermediate_supervision`

- `OPTIONAL`
- Type: `bool`
- Default: `true`
- Activate intermediate supervision
  - Intermediate supervision makes the loss computation more difficult, but improves the convergence by fixing the vanishing gradient issue
  - **USE OF INTERMEDIATE SUPERVISION HIGHLY RECOMMENDED**

`input_size`

- `OPTIONAL`
- Type: `int`
- Default: `256`
- The input size of the model

`output_size`

- `OPTIONAL`
- Type: `int`
- Default: `64`
- The heatmap size in model's output

`stages`

- `OPTIONAL`
- Type: `int`
- Default: `2`
- Number of [HourglassLayers](../hourglass_tensorflow/layers/hourglass.py) to use in the model

`downsampling_per_stage`

- `OPTIONAL`
- Type: `int`
- Default: `4`
- Number of downsampling operation in each [HourglassLayers](../hourglass_tensorflow/layers/hourglass.py)
  - The number of `downsampling_per_stage` must be equal to `input_size / output_size`

`stage_filters`

- `OPTIONAL`
- Type: `int`
- Default: `64`
- Number of latent space channels/filters within each[HourglassLayers](../hourglass_tensorflow/layers/hourglass.py)
  - The bigger the more computationally intensive

`output_channels`

- `OPTIONAL`
- Type: `int`
- Default: `64`
- Number of output channels/filters within each[HourglassLayers](../hourglass_tensorflow/layers/hourglass.py)
  - **MUST BE EQUAL TO JOINT NUMBER**

`name`

- `OPTIONAL`
- Type: `str`
- Default: `HourglassSample`
- Name of the model

## train

`object`

- `OPTIONAL`
- Type: [`HTFObjectReference[BaseTrainHandler]`](../hourglass_tensorflow/types/config/fields.py)
- Default: `hourglass_tensorflow.handlers.train.HTFTrainHandler`
- Reference to the handler to use for the Train process _(see [doc/HANDLERS](./HANDLERS.md))_

`epochs`

- `OPTIONAL (TRAIN)`
- Type: `int`
- Default: `10`
- Number of training epochs

`epoch_size`

- `OPTIONAL (TRAIN)`
- Type: `int`
- Default: `1000`
- Number of training iteration per epoch

`batch_size`

- `OPTIONAL (TRAIN)`
- Type: `int`
- Default: `128`
- Batch size

`optimizer`

- `OPTIONAL (TRAIN)`
- Type: `Union[str, HTFObjectReference[keras.Optimizer]]`
- Default: `keras.optimizers.RMSprop`
- Reference to the `keras.Optimizer` to use as optimizer _(see [doc/HANDLERS](./HANDLERS.md))_

`learning_rate`

- `OPTIONAL (TRAIN)`
- Type: `Union[float, HTFObjectReference[keras.LearningRateSchedule]]`
- Default: `keras.optimizers.schedules.learning_rate_schedule.ExponentialDecay`
- Reference to the `keras.LearningRateSchedule` to use as learning rate _(see [doc/HANDLERS](./HANDLERS.md))_
  - If set, `$train.optimizer` must be an `HTFObjectReference`

`loss`

- `OPTIONAL (TRAIN)`
- Type: `Union[str, HTFObjectReference[keras.Loss]]`

CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.1 (2022-08-26)
### Fix
* **metrics:** Avoid NaN metric value with safe divide. ([`ef58e54`](https://github.com/wbenbihi/hourglasstensorflow/commit/ef58e54ee75b36027184a3ba9ac4a64c642bba31))
* **metric:** PCK computation ([`57623c2`](https://github.com/wbenbihi/hourglasstensorflow/commit/57623c2507d3413320faa170bb7c2bff2c16dcce))

### Documentation
* Add badges for Publish and Test ([`30cc6a3`](https://github.com/wbenbihi/hourglasstensorflow/commit/30cc6a37a29cd77054cde9d89000dba2bf5c1a9d))

## v1.0.0 (2022-08-25)
### Feature
* **handlers:** Add Train Handler ([`a695850`](https://github.com/wbenbihi/hourglasstensorflow/commit/a69585036d73623d45add6ffc47b9e9f427cc1f7))
* **metrics:** Add OKS metric ([`d9c8ade`](https://github.com/wbenbihi/hourglasstensorflow/commit/d9c8ade014863e3b5bb9aa3ddcc6f9dbfee658a3))
* **metrics:** PCK metric ([`2013c7c`](https://github.com/wbenbihi/hourglasstensorflow/commit/2013c7cad1cbb0239073cdb4ec12bc2755550dc2))
* **metrics:** Add OverallMeanDistance metric ([`3424980`](https://github.com/wbenbihi/hourglasstensorflow/commit/3424980d865c2791504845156b650faefa187900))
* **func:** Add 2d argmax ([`3ca77bc`](https://github.com/wbenbihi/hourglasstensorflow/commit/3ca77bc603e76ff90f2b29b747fa5e81547f5df6))
* **loss:** Add SigmoidCrossEntropyLoss ([`7384c59`](https://github.com/wbenbihi/hourglasstensorflow/commit/7384c59cf407dec44f1850435205f656876e10c0))
* **cli:** Add summary command ([`7c975af`](https://github.com/wbenbihi/hourglasstensorflow/commit/7c975af263c4e07b69eeea59d8906f29f773a1b4))
* **cli:** Add plot model image ([`9fd9e19`](https://github.com/wbenbihi/hourglasstensorflow/commit/9fd9e19a4ddc65efbd74bd9b71889f76f06e54d1))
* **cli:** Add model cli commands ([`dbb3fc7`](https://github.com/wbenbihi/hourglasstensorflow/commit/dbb3fc7e3fd65190e6c1806de61617233bca5daf))
* ModelHandler ([`b6b96b5`](https://github.com/wbenbihi/hourglasstensorflow/commit/b6b96b5fc019b2aff4024f1d6e1474b1cc183e58))
* Handlers ([`28997b2`](https://github.com/wbenbihi/hourglasstensorflow/commit/28997b2d0d819ea222458a5fb7ccd7bddefbb93b))
* Layers for Hourglass Network ([`f23c7ad`](https://github.com/wbenbihi/hourglasstensorflow/commit/f23c7ad03daffa1d6ed63430d36a0855125e1a1c))
* Normalize data mapper in dataset handler ([`5a3c108`](https://github.com/wbenbihi/hourglasstensorflow/commit/5a3c108f5aec4380987e19fda1569034d55c31a9))
* Normalization method to config ([`1d85ff5`](https://github.com/wbenbihi/hourglasstensorflow/commit/1d85ff559bad8662c47339c5bd1368f36a48aed0))
* Allow additional params to init DatasetHandler ([`6723c8e`](https://github.com/wbenbihi/hourglasstensorflow/commit/6723c8ea7c5850b4830c1b79b9b5962bb567291d))
* Dataset Handler ([`a3d80a9`](https://github.com/wbenbihi/hourglasstensorflow/commit/a3d80a9c6d782642fe527b9bdf3ad43d43d86e93))
* Automated import from configuration ([`9df51a5`](https://github.com/wbenbihi/hourglasstensorflow/commit/9df51a5cbdf341d2f40d42194738f48455c75412))
* Tf.Dataset Handler ([`6cf74a1`](https://github.com/wbenbihi/hourglasstensorflow/commit/6cf74a1f1e6e9d89612f6e4ac0cf7f9cd2cac57b))
* Configuration Handler ([`ef22dfa`](https://github.com/wbenbihi/hourglasstensorflow/commit/ef22dfa9905570b91980482b2d9c9fa5c1d8f979))
* HTFPersonDatapoint plot functions ([`c00abd9`](https://github.com/wbenbihi/hourglasstensorflow/commit/c00abd98297c2c0def8d5afe9ce126e8d31ff0ab))
* Convert joints format in HTFPersonDatapoint ([`dcf4916`](https://github.com/wbenbihi/hourglasstensorflow/commit/dcf49162af4f200136457908ce5963248a5475be))
* Add convert from MPII to HTF cli command ([`d051841`](https://github.com/wbenbihi/hourglasstensorflow/commit/d0518414f0f112317c2392b55decb4a31e126616))
* Converter from MPII to HTF compliant data ([`33574d8`](https://github.com/wbenbihi/hourglasstensorflow/commit/33574d8f174031ca846900c7b931fee1b8c203ad))
* **cli:** Add CLI ([`b1796a9`](https://github.com/wbenbihi/hourglasstensorflow/commit/b1796a9693bdf8f76305a10b45875027b9196527))
* MPII Parsing ([`e6c84b8`](https://github.com/wbenbihi/hourglasstensorflow/commit/e6c84b88046dd537530bb4b5addf923ecaccc75f))

### Fix
* Bbox cordinates clamped to maximum size ([`fabe94a`](https://github.com/wbenbihi/hourglasstensorflow/commit/fabe94a103256b35986bceccd844facce4deeb5e))
* **metrics:** Cast metric results to float32 ([`bf57506`](https://github.com/wbenbihi/hourglasstensorflow/commit/bf5750646a5c671b86798c2af953505bb7904e06))
* **layers:** Super get_config methods for model serialization ([`5941db2`](https://github.com/wbenbihi/hourglasstensorflow/commit/5941db22a3000f361fea7c634e1fbfd6b099f408))
* **handlers:** Type hint after init_handler ([`809aabb`](https://github.com/wbenbihi/hourglasstensorflow/commit/809aabbc118fce95345071cdfc4b41ca9ea48331))
* **config:** Bad argument for TensorBoard callback ([`ec03a0d`](https://github.com/wbenbihi/hourglasstensorflow/commit/ec03a0d0354cfb361fe2d8d444ce5c95301424ee))
* **config:** Callback source from `tensorflow` ([`151b4dd`](https://github.com/wbenbihi/hourglasstensorflow/commit/151b4ddd0e2e800b18dbaa17162fb38657874265))
* **handlers:** Fix optimizer initialization with bad keyword lr ([`ee76786`](https://github.com/wbenbihi/hourglasstensorflow/commit/ee767866c76f9204c16a0e2b6876fd78010175a3))
* **loss:** Add default reduction in SigmoidCrossEntropyLoss ([`c3460a5`](https://github.com/wbenbihi/hourglasstensorflow/commit/c3460a5c39702f660b1ed42df8d01665321b2655))
* **config:** Bad source for ExponentialDecay ([`6c05364`](https://github.com/wbenbihi/hourglasstensorflow/commit/6c0536463b76277232baa2c008cf1b1dd3ed528a))
* **handlers:** Use map instead of apply for tf.data.Dataset ([`421d578`](https://github.com/wbenbihi/hourglasstensorflow/commit/421d578d4f53a17301f59719eda1ed76a6a37c21))
* **handlers:** Typo in _create_dataset method argument name ([`23a8a89`](https://github.com/wbenbihi/hourglasstensorflow/commit/23a8a89d79983dd74da6aaa5f0013409e925af7f))

hourglass_tensorflow/utils/config.py
import os
import json
import itertools
from glob import glob
from typing import TYPE_CHECKING
from typing import Set
from typing import Dict
from typing import List
from typing import Type
from typing import Union
from typing import Literal
from typing import Callable
from typing import Optional

import toml
import yaml
import pandas as pd
from loguru import logger
from pydantic import Field
from pydantic import BaseModel

from hourglass_tensorflow.utils._errors import BadConfigurationError
from hourglass_tensorflow.utils.object_logger import ObjectLogger
from hourglass_tensorflow.utils.parsers._parse_import import get_dataset

if TYPE_CHECKING:
    from hourglass_tensorflow.datasets import HTFBaseDatasetHandler


# region BaseModels


class HTFDatasetSplitConfig(BaseModel):
    column: str = "set"
    activate: bool = False
    train_value: str = "TRAIN"
    test_value: str = "TEST"
    validation_value: str = "VALIDATION"
    train_ratio: Optional[float] = 0.7
    validation_ratio: Optional[float] = 0.15
    test_ratio: Optional[float] = 0.15


class HTFDatasetSetsConfig(BaseModel):
    validation: bool = False
    test: bool = False


class HTFDatasetParamsConfig(BaseModel):
    class Config:
        extra = "allow"


class HTFColsConfig(BaseModel):
    cols: List[str] = Field(default_factory=list)


class HTFDatasetBBoxPaddingConfig(HTFColsConfig):
    x: int = 0
    y: int = 0


class HTFDatasetBBoxConfig(BaseModel):
    activate: bool = False
    factor: float = 1
    padding: HTFDatasetBBoxPaddingConfig = Field(
        default_factory=HTFDatasetBBoxPaddingConfig
    )


class HTFDatasetConfig(BaseModel):
    object: str
    input_size: int = 256
    output_size: int = 64
    stacks: int = 4
    heatmap_stddev: float = 10.0
    normalization: Optional[str] = None
    params: Optional[HTFDatasetParamsConfig] = Field(
        default_factory=HTFDatasetParamsConfig
    )
    sets: Optional[HTFDatasetSetsConfig]
    split: Optional[HTFDatasetSplitConfig] = Field(
        default_factory=HTFDatasetSplitConfig
    )
    bbox: Optional[HTFDatasetBBoxConfig] = Field(default_factory=HTFDatasetBBoxConfig)


ImageModes = Union[
    Literal["RGB"],
    Literal["rgb"],
    Literal["RGBA"],
    Literal["rgba"],
    Literal["GRAY"],
    Literal["gray"],
    Literal["GRAYSCALE"],
    Literal["grayscale"],
    Literal["BGR"],
    Literal["bgr"],
    Literal["BGRA"],
    Literal["bgra"],
]

CHANNELS_PER_MODE = {
    "RGB": 3,
    "rgb": 3,
    "RGBA": 4,
    "rgba": 4,
    "GRAY": 1,
    "gray": 1,
    "GRAYSCALE": 1,
    "grayscale": 1,
    "BGR": 3,
    "bgr": 3,
    "BGRA": 4,
    "bgra": 4,
}


class HTFDataInputConfig(BaseModel):
    mode: ImageModes = "RGB"
    source: str
    extensions: List[str] = Field(default_factory=["png", "jpeg", "jpg"])


class HTFDataOutputJointsFormatSuffixConfig(BaseModel):
    x: str = "X"
    y: str = "Y"

    class Config:
        extra = "allow"


class HTFDataOutputJointsFormatConfig(BaseModel):
    suffix: Optional[HTFDataOutputJointsFormatSuffixConfig] = Field(
        default_factory=HTFDataOutputJointsFormatSuffixConfig
    )


class HTFDataOutputJointsConfig(BaseModel):
    n: int = 16
    naming_convention: str = "joint_{JOINT_ID}_{SUFFIX}"
    format: Optional[HTFDataOutputJointsFormatConfig] = Field(
        default_factory=HTFDataOutputJointsFormatConfig
    )
    names: List[str] = Field(
        default_factory=[
            "00_rAnkle",
            "01_rKnee",
            "02_rHip",
            "03_lHip",
            "04_lKnee",
            "05_lAnkle",
            "06_pelvis",
            "07_thorax",
            "08_upperNeck",
            "09_topHead",
            "10_rWrist",
            "11_rElbow",
            "12_rShoulder",
            "13_lShoulder",
            "14_lElbow",
            "14_lWrist",
        ]
    )


class HTFDataOutputConfig(BaseModel):
    source: str
    source_column: str = "image"
    source_prefixed: bool = False
    prefix_columns: List[str] = Field(
        default_factory=[
            "set",
            "image",
            "scale",
            "bbox_tl_x",
            "bbox_tl_y",
            "bbox_br_x",
            "bbox_br_y",
            "center_x",
            "center_y",
        ]
    )
    set_column: str = "set"
    joints: HTFDataOutputJointsConfig


class HTFDataConfig(BaseModel):
    input: HTFDataInputConfig
    output: HTFDataOutputConfig


class HTFConfig(BaseModel):
    version: Optional[str]
    data: HTFDataConfig
    dataset: HTFDatasetConfig


# endregion

# region ConfigParser


class HTFConfigParser(ObjectLogger):
    """Parse configuration files for `hourglass_tensorflow`

    Currently HTFConfigParser supports `.json`, `.toml`, `.yaml`, `.yml` files.
    The configuration file must follow the model defined by `HTFConfig`,
    see `CONFIGURATION.md` for more details.

    Args:
        filename (str): Configuration file name
    """

    def __init__(self, filename: str, verbose: bool = True) -> None:
        """Init ConfigParser. See `help(HTFConfigParser)`"""
        self._verbose = verbose
        self._filename = filename
        self._data = {}
        self._config: Optional[HTFConfig] = None
        self.info("Reading configuration ...")
        # Parse Config File
        self._infer_source()
        self._parse_config()

    def __call__(self) -> HTFConfig:
        """Returns the parsed configuration as a `pydantic.BaseModel` object

        Returns:
            HTFConfig: The configuration object
        """
        return self._config

    def __repr__(self) -> str:
        return f"<HFTConfigParser type:{self.file_type} source:{self._filename}>"

    @property
    def config(self) -> HTFConfig:
        return self._config

    @property
    def file_type(self) -> HTFConfig:
        return self._filename.split(".")[-1]

    @classmethod
    def parse_toml(cls, filename: str) -> Dict:
        """Parse a `.toml` file

        Args:
            filename (str): toml file name

        Returns:
            Dict: parsed toml as `dict`
        """
        return toml.load(filename)

    @classmethod
    def parse_json(cls, filename: str) -> Dict:
        """Parse a `.json` file

        Args:
            filename (str): json file name

        Returns:
            Dict: parsed json as `dict`
        """
        with open(filename, "r") as f:
            return json.load(f)

    @classmethod
    def parse_yaml(cls, filename: str) -> Dict:
        """Parse a `.yml`/`.yaml` file

        Args:
            filename (str): yaml file name

        Returns:
            Dict: parsed yaml as `dict`
        """
        with open(filename, "r") as f:
            return yaml.safe_load(f)

    def _infer_source(self) -> None:
        parsers = {
            "yml": self.parse_yaml,
            "yaml": self.parse_yaml,
            "json": self.parse_json,
            "toml": self.parse_toml,
        }
        cfg_parser: Optional[Callable[[str], Dict]] = parsers.get(
            self._filename.split(".")[-1]
        )
        if cfg_parser is None:
            raise ValueError(
                "The file extension is not supported. Use YAML, TOML or JSON config file"
            )
        self._data = cfg_parser(self._filename)

    def _parse_config(self) -> None:
        self._config = HTFConfig.parse_obj(self._data)


# endregion

# region Configuration Handler


class HTFConfigMeta(BaseModel):
    available_images: Optional[Set[str]] = Field(default_factory=set)
    label_type: Optional[Union[Literal["json"], Literal["csv"]]]
    label_headers: Optional[List[str]] = Field(default_factory=list)
    label_mapper: Optional[Dict[str, int]] = Field(default_factory=dict)
    dataset_object: Optional[Type["HTFBaseDatasetHandler"]]

    class Config:
        extra = "allow"


class HTFConfiguration(ObjectLogger):
    def __init__(self, config_file: str, verbose: bool = True) -> None:
        self._verbose = verbose
        # Init data
        self._metadata = HTFConfigMeta()
        self._labels_df: pd.DataFrame = pd.DataFrame([])
        self.dataset_handler: Optional["HTFBaseDatasetHandler"] = None
        # Parse Configuration
        self._config = HTFConfigParser(filename=config_file)

    @property
    def config(self) -> HTFConfig:
        return self._config.config

    @property
    def _cfg_data(self) -> HTFDataConfig:
        return self.config.data

    @property
    def _cfg_data_inp(self) -> HTFDataInputConfig:
        return self.config.data.input

    @property
    def _cfg_data_out(self) -> HTFDataOutputConfig:
        return self.config.data.output

    # DATA - Prepare - Methods
    def _list_input_images(self) -> None:
        if not os.path.exists(self._cfg_data_inp.source):
            raise BadConfigurationError(
                f"Unable to find source folder {self._cfg_data_inp.source}"
            )
        self.info(
            f"Listing {self._cfg_data_inp.extensions} images in {self._cfg_data_inp.source}"
        )
        self._metadata.available_images = set(
            itertools.chain(
                *[
                    glob(os.path.join(self._cfg_data_inp.source, f"*.{ext}"))
                    for ext in self._cfg_data_inp.extensions
                ]
            )
        )

    def _valid_labels_header(self, df: pd.DataFrame, _error: bool = False) -> bool:
        # Check if numbers of columns are valid
        n_joint = self._cfg_data_out.joints.n
        # TODO(@wbenbihi) UNNECESSARY BLOCK
        # num_prefix_columns = len(self._cfg_data_out.prefix_columns)
        # num_columns_per_joint = len(
        #     self._cfg_data_out.joints.format.suffix.__fields__
        # )
        # estimated_column_size = n_joint * num_columns_per_joint + num_prefix_columns
        # if len(df.columns) != estimated_column_size:
        #     if _error:
        #         raise BadConfigurationError(
        #             "Labels Columns does not matched estimated columns"
        #         )
        #     return False
        # Check if columns names are valid
        naming_convention = self._cfg_data_out.joints.naming_convention
        suffixes = self._cfg_data_out.joints.format.suffix
        headers = self._cfg_data_out.prefix_columns + [
            naming_convention.format(JOINT_ID=jid, SUFFIX=suffix)
            for jid in range(n_joint)
            for suffix in suffixes.__dict__.values()
        ]
        if set(headers).difference(set(list(df.columns))):
            if _error:
                raise BadConfigurationError(
                    f"Columns' name does not match configuration\n\tEXPECTED:\n\t{headers}\n\tRECEIVED:\n\t{list(df.columns)}\n\tMISSING COLUMNS:\n\t{set(headers).difference(set(list(df.columns)))}"
                )
            return False
        # If everything is good we store the expected headers in _metadata
        self._metadata.label_headers = headers
        return True

    def _read_labels(self, _error: bool = False) -> bool:
        # Check if data.output.source exists ?
        if not os.path.exists(self._cfg_data_out.source):
            raise BadConfigurationError(f"Unable to find {self._cfg_data_out.source}")
        # Read Data
        self.info(f"Reading labels from {self._cfg_data_out.source}")
        ## Check if the file extension is in [.json, .csv]
        if self._cfg_data_out.source.endswith(".json"):
            self._metadata.label_type = "json"
            labels = pd.read_json(self._cfg_data_out.source, orient="records")
        elif self._cfg_data_out.source.endswith(".csv"):
            self._metadata.label_type = "csv"
            labels = pd.read_csv(self._cfg_data_out.source)
        else:
            raise BadConfigurationError(
                f"{self._cfg_data_out.source} should be of type .json or .csv"
            )
        if not isinstance(labels, pd.DataFrame):
            raise BadConfigurationError(
                f"{self._cfg_data_out.source} not parsable as pandas.DataFrame"
            )
        # Validate expected labels columns
        if not self._valid_labels_header(labels, _error=_error):
            self.error("Labels are not matching")
            return False
        self._labels_df: pd.DataFrame = labels[self._metadata.label_headers]
        self._metadata.label_mapper = {
            label: i for i, label in enumerate(self._metadata.label_headers)
        }
        if not self._cfg_data_out.source_prefixed:
            # Now we also prefix the image column with the image folder
            # in case the source_prefix attribute is set to false
            folder_prefix = self._cfg_data_inp.source
            source_column = self._cfg_data_out.source_column
            self._labels_df = self._labels_df.assign(
                **{
                    source_column: self._labels_df[source_column].apply(
                        lambda x: os.path.join(folder_prefix, x)
                    )
                }
            )
        return True

    def _prepare_input(self) -> None:
        # List files in Input Source Folder
        self._list_input_images()

    def _validate_joints(self, _error: bool = True) -> bool:
        conditions = [
            len(self._cfg_data_out.joints.names) == self._cfg_data_out.joints.n
        ]
        if not all(conditions):
            if _error:
                raise BadConfigurationError("Joints properties are not valid")
            return False
        return True

    def _prepare_output(self, _error: bool = True) -> None:
        # Read the label file
        self._validate_joints(_error=_error)
        self._read_labels(_error=_error)

    def prepare_data(self) -> None:
        self._prepare_input()
        self._prepare_output()

    def prepare(self) -> None:
        self.prepare_data()

    # DATASET - Prepare - Methods

    def _load_dataset_object(self) -> None:
        self._metadata.dataset_object = get_dataset(self.config.dataset.object)

    def prepare_dataset(self) -> None:
        self._load_dataset_object()
        self.dataset_handler: "HTFBaseDatasetHandler" = self._metadata.dataset_object(
            dataset=self._labels_df.to_numpy(),
            config=self.config.dataset,
            global_config=self,
            **self.config.dataset.params.dict(),
        )
        self.dataset_handler.execute()


# endregion
